### ISSUE TITLE
Spaces are not replaced with underscore (config setting) - forbidden blank page instead

### ISSUE THREAD

**Issue by macin** [dokuwiki/dokuwiki#4238]
### The problem

I have a config enabled to convert spaces to underscores.
However after upgrade to Kaos 6a I started getting feedback from the users, that they are not able to create pages. 

AFter investigation, is seems that when user types in the namespace name with spaces, dokuwiki redirects to the page with %20 in the url

I was not able to confirm if this is server problem, or dokuwiki itself

Enter page name here
![IMG_20240308_183310](https://github.com/dokuwiki/dokuwiki/assets/1422747/3b0f25f2-7948-4c30-8432-b084c4c9e89f)

![Zrzut ekranu 2024-03-08 174252](https://github.com/dokuwiki/dokuwiki/assets/1422747/63bb27f5-9a83-4fed-aa47-202e1bb14b82)

### Version of DokuWiki

2024-0206a Kaos

### PHP Version

8.2

### Webserver and version of webserver

Ubuntu

### Browser and version of browser, operating system running browser

Version 122.0.6261.71 (Official Build) (64-bit)

**Comment by fiwswe** [comment]
>AFter investigation, is seems that when user types in the namespace name with spaces, dokuwiki redirects to the page with %20 in the url

Where does the user type in the name with spaces? The URL bar of the browser?

If so then the browser most likely replaces the spaces with `%20` because space is not a valid character in URLs.

**Comment by fiwswe** [comment]
Yes, probably a plugin. DokuWiki does not have any buttons to create pages or modal dialogs to enter (new) page names.

Also, what template are you using? Might also be related?


>I have a config enabled to convert spaces to underscores.

Is this the [sepchar](https://www.dokuwiki.org/config:sepchar) setting? I have that set to `_` as well and never had any issues.

**Comment by fiwswe** [comment]
Not a DokuWiki issue. See: https://github.com/SoarinFerret/dokuwiki-plugin-pagebuttons/issues/20

**Comment by splitbrain** [comment]
I actually think there might be a regression here. Needs investigation.

**Comment by splitbrain** [comment]
Okay, it seems DokuWiki itself is still working as expected. If an ID with spaces is passed as ID, it will be cleaned up correctly: https://www.dokuwiki.org/doku.php?id=playground:page%20with%20spaces

However when the the ID is passed via mod_rewrite, a forbidden error is issued: https://www.dokuwiki.org/playground:page%20with%20spaces

 However I don't think it is triggered by DokuWiki but the web server? But I am not sure why... the rewrite should pass the ID on to doku.php, space or no space?

If @macin is correct and this used to work in Jack Jackrum, I have no idea why. Or what the change is that causes the regression. Would be great if someone could confirm that it works on Jack.

**Comment by fiwswe** [comment]
Ok, I (also?) tried this and to my surprise it worked:
https://www.dokuwiki.org/playground:page_with_spaces

The cleanup of the page id happens when saving the page. While initially editing it, it still had the spaces/`%20` in the URL.

I can't say what happens in the mod_rewrite case.

Depends on the exact rewrite rules. If the rules from https://www.dokuwiki.org/install:apache are used then the URL would start out as: `https://hostname/ns:page%20with%20spaces&do=edit`
Thus the `QSD` modifier would not see a query string (which would need to start with `?`) and the `B` modifier would escape the `&` -> `%26` and `=` -> `%3D` yielding: `https://hostname/doku.php?id=ns:page%20with%20spaces%26do%3Dedit`. Would DokuWiki consider this page name forbidden? I tried https://www.dokuwiki.org/playground:page%20with%20spaces$26do%3Dedit and it does indeed return Forbidden.

**Comment by boutmos** [comment]
Got the same issue on Jack Jackrum. Fixed %2F and %20 in master/lib/plugins/indexmenu/scripts/contextmenu.js on function getid :

```
    /**
     * Build a url to a wiki page
     *
     * @param {string} urlbase
     * @param {string} id
     * @returns {string}
     */
    getid: function (urlbase, id) {
      // Replace spaces with + before encoding
      id = id.replace(/ /g, '+');
      // Split the id by slashes, encode each part, and join them back with slashes
      let encodedId = id.split('/').map(encodeURIComponent).join('/');
      let url = (urlbase || '') + encodedId;
      url += (urlbase.indexOf("?") < 0) ? '?' : '&';
      return url;
    },
```

in local.protected.php put also $conf['sepchar'] = '-';

**Comment by fiwswe** [comment]
@boutmos Well then, even if your observations seem to be similar, that would still make them an a problem with the [IndexMenu Plugin](https://www.dokuwiki.org/plugin:indexmenu). I don't see any issue at the DokuWiki level. And the discussion about the original problem pointed at the [Page Buttons Plugin](https://www.dokuwiki.org/plugin:pagebuttons) and an interaction with certain optional [Apache httpd](https://httpd.apache.org) configuration settings.

You should file an issue for the [IndexMenu Plugin](https://www.dokuwiki.org/plugin:indexmenu) at https://github.com/samuelet/indexmenu.

**Comment by macin** [comment]
Ok, partial success. Adding the B flag worked on my local env based on dokuwikistick server. However still not working with the production server.

**Comment by splitbrain** [comment]
Which Apache version are you running in production? Is there any other server in front? Like a reverse proxy?

**Comment by macin** [comment]
Ok, we looked inside with devops guys, and we found that the htaccess was being overwritten by the some apache config. So one hurdle overcome

The apache version is: Apache/2.4.57 (Debian)

![2024-11-26_09h42_09](https://github.com/user-attachments/assets/af9956d1-bae8-4067-b7e5-c1e33fd1ff23)

we managed to get to this point (I'd expected to see already an editor) 
Nevertheless, clicking the button to create page does not work either, so still something is not correct... no ideas what to go for next...
